### PR TITLE
feat: add expressionUseOnlyWhitelistedLabelsForMetric

### PR DESCRIPTION
**Pull request by david-vavra**
Adding a new validation - expressionUseOnlyWhitelistedLabelsForMetric :)

### CODE REVIEW

File: docs/validations.md [hunk 1]
```diff
@@ -226,6 +227,18 @@ params:
...
 
+#### `expressionUseOnlyWhitelistedLabelsForMetric`
+
+Fails if the rule uses any labels beside those provided in combination with given metric in its `expr` label matchers, aggregations or joins.
+> If using kube-state-metrics for exposing labels information about K8S objects (kube_*_labels) only those labels whitelisted by kube-state-metrics admin will be available.
+> Might be useful to check that users does not use any other in their expressions.
+
+```yaml
+params:
+  metric: "kube_pod_labels"
```
Comment (david-vavra) [inline docs/validations.md:238]: I was thinking what metric could be regexp, wdyt?

**Comment by FUSAKLA** [comment]
Also, please rebase since I merged some dependency updates and bumped version if linting in CI :pray:

File: pkg/validator/expression.go [hunk 2]
```diff
@@ -104,6 +104,47 @@ func (h expressionDoesNotUseLabels) Validate(_ unmarshaler.RuleGroup, rule rulef
...
+		return []error{err}
+	}
+	allowedLabelsMap := map[string]struct{}{}
+	for _, l := range h.labels {
+		allowedLabelsMap[l] = struct{}{}
+	}
+	var errs []error
+	for _, l := range usedLabels {
+		if _, ok := allowedLabelsMap[l]; !ok {
+			errs = append(errs, fmt.Errorf("forbidden label `%s` used in expression", l))
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply pkg/validator/expression.go]: Changed to ```forbidden label `%s` used in expression in combination with metric %s (regexp)```

File: pkg/validator/expression_helpers.go [hunk 3]
```diff
@@ -34,6 +34,63 @@ func getExpressionUsedLabels(expr string) ([]string, error) {
 	return slices.Compact(usedLabels), nil
 }
 
+// Returns true in case selector matches given metric and a list of other labels beside __name__, false and nil otherwise.
+func containsMetric(selector *parser.VectorSelector, metric string) (metricUsed bool, usedLabels []string) {
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply pkg/validator/expression_helpers.go]: changed as suggested

File: pkg/validator/expression_helpers.go [hunk 4]
```diff
@@ -34,6 +34,63 @@ func getExpressionUsedLabels(expr string) ([]string, error) {
...
+					usedLabels = append(usedLabels, v.VectorMatching.MatchingLabels...)
+				}
+			}
+		}
+		return usedLabels
+	}
+
+	parser.Inspect(promQl, func(n parser.Node, path []parser.Node) error {
+		if v, isVectorSelector := n.(*parser.VectorSelector); isVectorSelector {
+			containsMetric, selectorUsedLabels := containsMetric(v, metric)
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply pkg/validator/expression_helpers.go]: Changed to avoid name colission

File: docs/validations.md [hunk 5]
```diff
@@ -226,6 +227,18 @@ params:
   labels: [ "foo", "bar" ]
 ```
 
+#### `expressionUseOnlyWhitelistedLabelsForMetric`
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply docs/validations.md]: Changed to `expressionUsesOnlyAllowedLabelsForMetricRegexp`

File: pkg/validator/expression_helpers.go [hunk 6]
```diff
@@ -34,6 +34,63 @@ func getExpressionUsedLabels(expr string) ([]string, error) {
 	return slices.Compact(usedLabels), nil
 }
 
+// Returns true in case selector matches given metric and a list of other labels beside __name__, false and nil otherwise.
+func containsMetric(selector *parser.VectorSelector, metric string) (metricUsed bool, usedLabels []string) {
+	for _, m := range selector.LabelMatchers {
+		if m.Name == "__name__" && m.Type == labels.MatchEqual && m.Value == metric {
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply pkg/validator/expression_helpers.go]: Changed, now metricNameRegexp is matched against both cases, either equal match or regexp (positive) patch.

**Comment by david-vavra** [comment]
All suggestions should be implemented now.

I'm not sure about the scenario in case validated expr will contain regexp match on metric name (e.g. `{__name__=~"kube_.*_.labels"}`).  Should it then be compared using regexp match with the configured metricNameRegexp, what in case this will be an actual regexp as well?

As it is now, the validation may be conterintuitive:
https://github.com/FUSAKLA/promruval/blob/836a2af170629b72ece4a37d7c262a63d77b2f49/pkg/validator/validator_test.go#L94

**Comment by FUSAKLA** [comment]
Great, thanks! 

Regarding the issue with a metric name selector being a regexp, the correct solution would be to tell if both the regexps intersect in its matches, but that's not a simple task AFAIK.

Looking at the possibilities of the go regexp package, I'd go with 

1. Get [LiteralPrefix](https://pkg.go.dev/regexp#Regexp.LiteralPrefix) of the `metricNameRegexp` 
  - If it returns `complete==true` (means the `mtericNameRegexp` wasjust plain string) -> try if the `__name__` regexp matches the returned literal prefix
  - if `complete==false`, try if the `metricNameRegexp` matches the LiteralPrefix of the `__name__` regexp
  
  Does this sound somehow reasonable? :thinking: 
  
  It's not perfect, but I cannot think of anything better than this "heuristic", but I'd say it's enough

**Comment by david-vavra** [comment]
> Great, thanks!
> 
> Regarding the issue with a metric name selector being a regexp, the correct solution would be to tell if both the regexps intersect in its matches, but that's not a simple task AFAIK.
> 
> Looking at the possibilities of the go regexp package, I'd go with
> 
> 1. Get [LiteralPrefix](https://pkg.go.dev/regexp#Regexp.LiteralPrefix) of the `metricNameRegexp`
> 
> * If it returns `complete==true` (means the `mtericNameRegexp` wasjust plain string) -> try if the `__name__` regexp matches the returned literal prefix

Sounds good.


> * if `complete==false`, try if the `metricNameRegexp` matches the LiteralPrefix of the `__name__` regexp

I did not want to use such heuristic from the start, as it may return false-positives, for example:

```
params:
  metricNameRegexp: "kube_.*_labels"
  allowedLabels: [ "pod", "cluster", "app", "team" ]
```
```
record:
  expr: {__name__=~"kube_.*_annotations"...} 
```

This is rather an edge-case and cannot even think about any real use-case with expr which would use regexp match for `__name__`. But I guess this can be left up for a user to consider as long as it will be documented (it has to be :).

**Comment by FUSAKLA** [comment]
You're about the false positives :/
In such case yes, lets rather not check the regexp __name__ matchers at all and just add warning :+1:

**Comment by david-vavra** [comment]
> You're about the false positives :/ In such case yes, lets rather not check the regexp **name** matchers at all and just add warning 👍

Done, PTAL. :)

File: docs/validations.md [hunk 7]
```diff
@@ -257,9 +258,26 @@ params:
...
+```yaml
+params:
+  metricNameRegexp: "kube_pod_labels"
+  allowedLabels: [ "pod", "cluster", "app", "team" ]
+```
+
 #### `expressionDoesNotUseOlderDataThan`
 
-Fails if the rule `expr` uses older data than specified limit in Prometheus duration syntax. Checks even in sub-queries
+Fails if the rule `expr` uses older data than specified limit in Prometheus duration syntax. Checks even in sub-queriesmetric
```
Comment (FUSAKLA) [inline docs/validations.md:280]: This change seems unrelated? :thinking: 

```suggestion
Fails if the rule `expr` uses older data than specified limit in Prometheus duration syntax. Checks even in sub-queries
```

File: pkg/validator/promql_expression.go [hunk 8]
```diff
@@ -121,6 +121,47 @@ func (h expressionDoesNotUseLabels) Validate(_ unmarshaler.RuleGroup, rule rulef
...
+		AllowedLabels    []string `yaml:"allowedLabels"`
+		MetricNameRegexp string   `yaml:"metricNameRegexp"`
+	}{}
+	if err := paramsConfig.Decode(&params); err != nil {
+		return nil, err
+	}
+	if len(params.AllowedLabels) == 0 {
+		return nil, fmt.Errorf("missing labels")
+	}
+	return &expressionUsesOnlyAllowedLabelsForMetricRegexp{allowedLabels: params.AllowedLabels, metricNameRegexp: params.MetricNameRegexp}, nil
```
Comment (FUSAKLA) [inline pkg/validator/promql_expression.go:140]: I completely missed this first time, but please make the `expressionUsesOnlyAllowedLabelsForMetricRegexp.metricNameRegexp` a `regexp.Regexp` type and compile it in the `newExpressionUseOnlyWhitelistedLabelsForMetric` so invalid regexp is caught during unmarshalling? This wait it could result in panic because of user input.

Same as here for example https://github.com/FUSAKLA/promruval/blob/414a2fc6008248c17fdef794f0d0f3f218a0f1ff/pkg/validator/others.go#L20

Also make the regexp anchored already in the `newExpressionUseOnlyWhitelistedLabelsForMetric` function and I see the `regexp.Compile("^" + regexp + "$")` in multiple places ([here](https://github.com/FUSAKLA/promruval/blob/414a2fc6008248c17fdef794f0d0f3f218a0f1ff/pkg/validator/promql_expression.go#L456) [here](https://github.com/FUSAKLA/promruval/blob/414a2fc6008248c17fdef794f0d0f3f218a0f1ff/pkg/validator/others.go#L20)). Might be worth add a `compileAnchoredRegexp(string) (regexp.Regexp, error)` function?

File: pkg/validator/promql_expression_helpers.go [hunk 9]
```diff
@@ -34,6 +35,63 @@ func getExpressionUsedLabels(expr string) ([]string, error) {
 	return slices.Compact(usedLabels), nil
 }
 
+// Returns true in case metric name selector matches given regexp and a list of other labels beside __name__, false and empty list otherwise.
+func labelsUsedInSelectorForMetricRegexp(selector *parser.VectorSelector, metricRegexp string) (usedLabels []string, metricUsed bool) {
```
Comment (FUSAKLA) [inline pkg/validator/promql_expression_helpers.go:39]: Compiling is out of scope of the function and if not, the error should be returned
```suggestion
func labelsUsedInSelectorForMetricRegexp(selector *parser.VectorSelector, metricRegexp *regexp.Regexp) (usedLabels []string, metricUsed bool) {
```

File: pkg/validator/promql_expression.go [hunk 10]
```diff
@@ -121,6 +121,47 @@ func (h expressionDoesNotUseLabels) Validate(_ unmarshaler.RuleGroup, rule rulef
...
+func (h expressionUsesOnlyAllowedLabelsForMetricRegexp) String() string {
+	return fmt.Sprintf("expression only uses allowed labels `%s` for metric %s", strings.Join(h.allowedLabels, "`,`"), h.metricNameRegexp)
+}
+
+func (h expressionUsesOnlyAllowedLabelsForMetricRegexp) Validate(_ unmarshaler.RuleGroup, rule rulefmt.Rule, _ *prometheus.Client) []error {
+	usedLabels, err := getExpressionUsedLabelsForMetricRegexp(rule.Expr, h.metricNameRegexp)
+	if err != nil {
+		return []error{err}
+	}
+	allowedLabelsMap := map[string]struct{}{}
```
Comment (FUSAKLA) [inline pkg/validator/promql_expression.go:152]: What about making already the `expressionUsesOnlyAllowedLabelsForMetricRegexp.allowedLabels` a `map[string]struct{}{}` so this happens only once during config unmarshalling and not on every validation?

File: pkg/validator/promql_expression_helpers.go [hunk 11]
```diff
@@ -34,6 +35,63 @@ func getExpressionUsedLabels(expr string) ([]string, error) {
...
+		}
+		usedLabels = append(usedLabels, m.Name)
+	}
+	return usedLabels, metricUsed
+}
+
+// Returns a list of labels which are used in given expr in relation to given metric.
+// Beside labels within vector selector itself, it adds labels used in Aggregate expressions and labels used in Binary expression.
+// For Binary expressions it may report false positives as the current implementation does not consider on which side of group_left/group_right is the given metric.
+func getExpressionUsedLabelsForMetricRegexp(expr, metricRegexp string) ([]string, error) {
```
Comment (FUSAKLA) [inline pkg/validator/promql_expression_helpers.go:53]: This whole function looks like a lot of logic duplication from the `getExpressionUsedLabels` above

Maybe adding a new `getExpressionUsedLabelsForMetric(expr, regexp)` and making `getExpressionUsedLabels` just wrapper for `getExpressionUsedLabelsForMetric(expr, ".*")` could suffice?

File: docs/validations.md [hunk 12]
```diff
@@ -257,9 +258,26 @@ params:
...
+* `{__name__="foo",..}, foo{...}` - `foo` is matched literally against given `metricNameRegexp`, if matches, expr is validated against `allowedLabels`
+* `{__name__=~"foo",..}` - skipped
+* `{__name__!="foo",}, {__name__!~"foo"}` - skipped
+
+> If using kube-state-metrics for exposing labels information about K8S objects (kube_*_labels) only those labels whitelisted by kube-state-metrics admin will be available.
+> Might be useful to check that users does not use any other in their expressions.
+
+```yaml
+params:
+  metricNameRegexp: "kube_pod_labels"
```
Comment (FUSAKLA) [inline docs/validations.md:274]: Other validators [do state this](https://github.com/FUSAKLA/promruval/blob/48590e41f6db1715f3bcc01498915449ab259b69/docs/validations.md#hassourcetenantsformetrics) so to be consistent
```suggestion
  metricNameRegexp: "kube_pod_labels" # The regexp will be fully anchored (surrounded by ^...$)
```

File: docs/validations.md (hunk 7 above)
Comment (david-vavra) [reply docs/validations.md:280]: mea culpa

File: pkg/validator/promql_expression.go (hunk 8 above)
Comment (david-vavra) [reply pkg/validator/promql_expression.go:140]: Refactored as proposed.

File: pkg/validator/promql_expression.go [hunk 13]
```diff
@@ -121,6 +121,47 @@ func (h expressionDoesNotUseLabels) Validate(_ unmarshaler.RuleGroup, rule rulef
...
+		return nil, err
+	}
+	if len(params.AllowedLabels) == 0 {
+		return nil, fmt.Errorf("missing labels")
+	}
+	return &expressionUsesOnlyAllowedLabelsForMetricRegexp{allowedLabels: params.AllowedLabels, metricNameRegexp: params.MetricNameRegexp}, nil
+}
+
+func (h expressionUsesOnlyAllowedLabelsForMetricRegexp) String() string {
+	return fmt.Sprintf("expression only uses allowed labels `%s` for metric %s", strings.Join(h.allowedLabels, "`,`"), h.metricNameRegexp)
```
[the remark this answers is not in the document]
Comment (david-vavra) [reply pkg/validator/promql_expression.go:144]: Changed as proposed.

File: pkg/validator/promql_expression_helpers.go (hunk 9 above)
Comment (david-vavra) [reply pkg/validator/promql_expression_helpers.go:39]: https://github.com/FUSAKLA/promruval/pull/70#discussion_r1712989880

File: pkg/validator/promql_expression.go (hunk 10 above)
Comment (david-vavra) [reply pkg/validator/promql_expression.go:152]: Refactored as proposed.

File: pkg/validator/promql_expression_helpers.go (hunk 11 above)
Comment (david-vavra) [reply pkg/validator/promql_expression_helpers.go]: Those two functions seemed to me different, as their approach to going through AST is not the same, but I guess you are right, and really `getExpressionUsedLabels` can be seen as a special case of the other one.

Refactored as suggested.

**Review by FUSAKLA — APPROVED** [review]
Nice! :+1: 

Thanks for addressing all the comments, and sorry for missing them during the initial review.
